Fix groups_controller

noriyotcp opened the pull request:
~~adminが一人の時のみgroupを削除できるようにした~~

編集画面に削除ボタンを追加

![image](https://user-images.githubusercontent.com/5820754/44977586-65f6c000-afa3-11e8-819b-e6a84f865d0b.png)


重要な操作なので 一番下にした。スタイルは `Save` buttonに合わせている
ボタンのサイズ、マージン共に大きいので浮いた感じになったが、重要なのでむしろ浮いていたほうがいいのでは、と思いました

（メモ転記）
- [x] Project ownerになっているGroupは削除不可
- [x] Group#membersが空のGroupは生まれないようにする
- [x] Group#membersにMembership#roleがadminなUserが最低一人いるようにする
  - #56 にて対応

takeyuweb commented on app/controllers/groups_controller.rb:
```diff
@@ -38,8 +38,12 @@ def update
-    @group.destroy
-    redirect_to :groups
+    if @group.admins.one?
+      @group.destroy
```
@noriyotcp (@itkrt2y ) プロジェクトのオーナーになったりしてますが、削除 `destroy` して大丈夫ですかね。

itkrt2y replied:
GroupがOwnerなものの後処理もしないとダメですね

itkrt2y replied:
Groupに関係ありそうなのだと `Project` , `Collaboration` , `Membership` がある

itkrt2y commented on app/controllers/groups_controller.rb:
```diff
@@ -11,10 +10,11 @@ def index
 
   def create
-    @group = Group.new group_params
+    @group = Group.new(group_params)
```
そもそも今のコードが問題な奴ですが、actionに入ってからインスタンス変数に入れた場合、 `authorize_resource` が無意味になりますね。
僕的には `authorize_resource` を消して、アクションごとに `can?(:create, @group)` とやる方が好き。

itkrt2y commented:
>adminが一人の時のみgroupを削除できるようにした

これは違いますね

Groupに何人所属していようが、admin権限を持つ人ならグループを一気に消せるべき。
ここにこんな制限を課しても、Group削除時に全員取り除く手間が増えるだけで特にメリットはない。むしろ、将来「全ユーザーremoveボタンをつける」って仕様が必要になる可能性を孕んでいて逆効果ですらある。

Groupと、User/Groupの紐付けであるMembershipの削除をしたいだけなのに、話を難しくしすぎです

itkrt2y commented:
>adminが一人の時のみgroupを削除できるようにした

これは違いますね

Groupに何人所属していようが、admin権限を持つ人ならグループを一気に消せるべき。
ここにこんな制限を課しても、Group削除時に全員取り除く手間が増えるだけで特にメリットはない。むしろ、将来「全ユーザーremoveボタンをつける」って仕様が必要になる可能性を孕んでいて逆効果ですらある。

Groupと、User/Groupの紐付けであるMembershipの削除をすればいいだけ

itkrt2y commented:
~Groupへの制限は必要なくてGroupとGroupに紐づくものを全部消せばいいだけですけど、Membershipには制限が必要で、adminは最低一人いるようにしないといけない。ここだけは注意。~

間違えてました。以下のメモを参考にしてください

noriyotcp committed:
Add is_deleted to Group and soft_destroy!

groupに紐づいたcollaborationsは全削除する

groups/editにdeleteボタンを設置

Remove authorize_resource
代わりにforbiddenを追加
- 権限がないユーザーはアクセス拒否

Add Group#deletable?
- Not delete a group when it is owner of project

itkrt2y commented:
メモ

- Project ownerになっているGroupは削除不可
- Group#membersが空のGroupは生まれないようにする
- Group#membersにMembership#roleがadminなUserが最低一人いるようにする

noriyotcp commented on app/controllers/groups_controller.rb:
```diff
@@ -54,11 +62,13 @@ def additional_params
 
-    def load_group
-      @group = Group.friendly.find params[:id]
+    def access_denied
```
`application_controller` の`CanCan::AccessDenied` を模倣したが、権限の無いユーザーが編集画面にアクセスした時にこうなる（変更前もこうなるのかは未確認）

![image](https://user-images.githubusercontent.com/5820754/45105065-7018ea00-b16e-11e8-8831-d20e10fff453.png)

これはかっこ悪いのでなんとかせねば
そもそも`Test01Group` のリンクが`edit` へのリンクになっているので、権限の無い場合はグループ詳細へのリンクにする、とか

itkrt2y replied:
401は認証していない場合で、リソースの変更が行えない場合は403 Forbiddenでしょうね

itkrt2y commented on app/controllers/groups_controller.rb:
```diff
@@ -30,16 +29,25 @@ def edit
+
+    projects = @group.projects.where(is_deleted: false)
+    projects.each { |p| p.update!(is_deleted: true) }
+    @group.destroy!
```
複数のUPDATEを行うのであれば、それらはtransactionで囲む。また、それをModelに書く

itkrt2y commented on app/models/group.rb:
```diff
@@ -46,6 +46,10 @@ def generate_draft
   end
 
+  def deletable?
+    projects.none?
```
今の実装、以下の流れになってて、意味わかんないですね。

1. Groupに紐付くProjectがない場合
1. Groupに紐付くProjectの削除フラグを立てて（← 紐付くProjectがないのに？）
1. Groupを削除する

あと先日コメントに書きましたけど、GroupもUserと同じでProject ownerになっている場合は削除不可能。
Projectは論理削除で生き残るので、紐付くownerがいなくなったら困る。（いなくてもいいように実装してもいいけど、nullableなのは面倒の方が多そう）

Groupも論理削除可能にするといいのでは

noriyotcp replied:
すみません、確認させてください
`destroy` にてprojectsがあるものは削除できないようにしているので、
残るは「projectsを持っていないグループ」ということになり、紐づいているprojectsがないのなら、グループも削除できるのではないでしょうか？

これも私の認識が間違っているのかどうかわからないです 😩

itkrt2y replied:
紐づいてるprojectがなければgroupを削除できるのはそうですが、話の流れ及びロジックとしては全く違いますね
その説明は、以下のロジックの説明には全くなっていませんし
> 1. Groupに紐付くProjectがない場合
> 2. Groupに紐付くProjectの削除フラグを立てて（← 紐付くProjectがないのに？）

結論を言うべきところですが僕が結論を言わないと良い設計が出てこないのも難なのでまずは質問。
なんでprojectがあるものは削除できないんですか？ 別の考え方をすると、projectがある場合はgroupを削除できない場合、groupを消したいユーザーはどのように行動しないといけませんか？

noriyotcp replied:
> projectがある場合はgroupを削除できない場合、groupを消したいユーザーはどのように行動しないといけませんか？

groupに紐づいているprojectsを全部削除しないといけないですね
うーん、でもprojectは論理削除なのか 🤔

itkrt2y replied:
僕がこのスレッド一番上のコメントを「Groupも論理削除可能にするといいのでは」で締めくくっている意味を考えてみてください

noriyotcp replied:
projectを論理削除(is_deleted: true)する
⬇️ 
`@group.destroy`する
⬇️ 
Project#owner_idはnot null
⬇️ 
owner(group) がいないので
＿人人人人人人＿
＞　突然の死　＜
￣Y^Y^Y^Y^Y^Y^￣

itkrt2y replied:
>groupに紐づいているprojectsを全部削除しないといけないですね

ここまでは考えられているのに、「どうせユーザーにGroup削除前にProject全て削除させるなら、システムでGroup論理削除時に一緒にProjectも論理削除すればいいじゃん」ってならないのが惜しいですね

itkrt2y replied:
「Project ownerになっているGroupは削除不可」って条件は物理削除してprojects.owner_idがnullになるのはまずいって話で、論理削除ならば無関係

itkrt2y commented on app/views/groups/_form.html.slim:
```diff
@@ -1,5 +1,6 @@
 = form_for @group, class: "group-form" do |f|
-  p#notice = notice
+  - flash.each do |key, value|
+    = content_tag(:p, value, id: "#{key}")
```
なぜkeyを式展開にしているのでしょう？

itkrt2y commented on app/models/group.rb:
```diff
@@ -46,6 +46,15 @@ def generate_draft
+    transaction do
+      update!(is_deleted: true)
+      collaborations.destroy_all
+      memberships.destroy_all
```
どのような理由からcollaborationとmembershipを削除することにしましたか？（理由があるのならば消すのはいいのですが、僕なら何もしない）

noriyotcp replied:
単にGroupに関連するものは消そうと思ってやったのですが、
Groupが論理削除になって実際には消えてないのに、これらを物理削除するのは不自然、という認識でしょうか？

itkrt2y replied:
>Groupが論理削除になって実際には消えてないのに、これらを物理削除するのは不自然、という認識でしょうか？

秋田さんはこれを自然だと思っていますか？ それとも不自然だと思っていますか？
また、どちらの答えにしてもそう考える根拠はなんですか？

noriyotcp replied:
繰り返しますが、単にGroupを削除したので関連も削除していたのですが、
論理削除になったので、関連も削除するのは不自然ということかな、と思って質問しただけです

itkrt2y replied:
自分で考えた結論をまずは断言して欲しいのです

itkrt2y replied:
根拠のない当てずっぽうを僕がOKというまで繰り返されても困りますし、ちゃんと考えて根拠を言語化できるようになって欲しいので

itkrt2y replied:
でもまあ一つ質問。論理削除のメリットってなんですか？

noriyotcp replied:
後から復活できるってことですかね
その時に関連がすっかり無くなっているのは不自然というかユーザーが困るというか

itkrt2y replied:
そうですね。今回はあまり関係ないですが、論理削除の一般的なメリットは可逆性です。
物理削除をすると（バックアップという手はあるにしろ）レコードを元通りにできない不可逆性がありますが、論理削除の場合は削除フラグを消すだけで復活できるのがメリットです。
その可逆性がメリットの機能を使うのにわざわざ中で不可逆なことをするとメリットが死にますけど、それにはどんな理由があるのですかと質問しているのがこのスレッド1つ目のコメント。

itkrt2y replied:
で、メリットを理解しちゃんと考えてるならば論理削除に物理削除は混ぜないと僕は思うのですが、行き当たりばったりに物理削除したり論理削除したり消さなくていいコードを消してみたりするもんだから、ただ答えを言うだけじゃいつまで経っても終わらないと思って回りくどいですけど質問してみてるのがこのスレッド

itkrt2y commented on spec/models/group_spec.rb:
```diff
@@ -30,4 +30,23 @@
+
+    it { expect { subject }.to change { group.is_deleted }.from(false).to(true) }
+    it { expect { subject }.to change { group.collaborations.count }.to(0) }
+    it { expect { subject }.to change { group.memberships.count }.to(0) }
```
`and`で繋げばいいかな

```rb
expect{ subject }.to change{ group.is_deleted }.from(false).to(true)
                .and change{ group.collaborations.count }.by(-2)
                .and change{ group.memberships.count }.by(-2)
```

itkrt2y commented on app/models/group.rb:
```diff
@@ -46,6 +46,15 @@ def generate_draft
+      update!(is_deleted: true)
+      collaborations.destroy_all
+      memberships.destroy_all
+      projects.where(is_deleted: false).each { |p| p.update!(is_deleted: true) }
```
削除フラグ立てるのにバリデーション通す必要ないでしょうし、 `projects.where(is_deleted: false).update_all(is_deleted: true)` でいいですよ

itkrt2y replied:
transactionの仕様をまだ完全に理解してないけど、 `update_all` だとrollbackされなかったりするだろうか？
バリデーションしないならARレイヤであげるエラーはないだろうし、transactionはRDB側の機能でRDB側で `UPDATE` 中になんらかの問題があったらrollbackされると思うが......

itkrt2y commented on app/models/group.rb:
```diff
@@ -26,8 +27,7 @@ class Group < ApplicationRecord
   friendly_id :name, use: :slugged
-
-  has_many :memberships, dependent: :destroy
+  has_many :memberships
```
`dependent: :destroy` であることには変わりないのではないですか？
実際に物理削除することはないにしても。

noriyotcp replied:
物理削除しないから不要かと思いましたが、確かに関係性は変わりないですね、なるほど

noriyotcp committed:
Remove Group#deletable?

groupを削除した場合:
group, projectsの論理削除

itkrt2y commented on app/models/group.rb:
```diff
@@ -47,15 +47,9 @@ def generate_draft
-      collaborations.destroy_all
-    end
+    update!(is_deleted: true)
+    projects.where(is_deleted: false).update_all(is_deleted: true)
```
`transaction` で囲むのは必要ですよ？（実際に活用されることはないと思いますが）
`update_all`  にするっていうのはProjectのvalidationを通さなくていいって話でこの2行がatomicな処理であることには変わりなくて、例えばDBで何かエラーが起きたりしたらrollbackされて欲しいのは変わりないので、 SQLレベルでtransactionで囲むためにRailsレイヤで `transaction` ブロックで囲むのは変わらない。

noriyotcp replied:
transaction外してもテストが通ってるのでロールバックされてるのかと思いきや、単に1行目でコケているので2行目が実行されていないというだけでした 😩

noriyotcp committed:
Do not display deleted groups

noriyotcp commented on app/views/layouts/_setting_links.html.slim:
```diff
@@ -13,7 +13,7 @@
 
     ul.groups
-      - current_user.groups.each do |group|
+      - current_user.groups.active.each do |group|
```
ユーザー管理画面左、削除されたグループを表示しないようにした

before
![_setting_links-in-groups-before](https://user-images.githubusercontent.com/5820754/45883802-14e12b80-bded-11e8-971d-c12d5f6c4a20.png)

after
![_setting_links-in-groups-after](https://user-images.githubusercontent.com/5820754/45883825-20cced80-bded-11e8-88de-2fac61755d50.png)

itkrt2y replied:
揚げ足取るようですが、論理削除は新機能なのでbeforeはないと思いますが......
「削除されたグループを表示しないようにした」って表現では、まるで表示されてはいけないものが表示されるバグがあったみたい

noriyotcp replied:
比較の為に変更「前」と「後」を提示しただけです
で、表示しないようにしたのだから、その事実を書いただけです
バグだったのならその旨を説明すればいいですし

itkrt2y replied:
新機能を前提にした修正なのだから比較対象が存在しない、って話です

itkrt2y replied:
比較対象が存在しないのにまるで存在するような表現をしているから、既存バグの説明をしているように見える、ということ

noriyotcp commented on app/views/owners/show.html.slim:
```diff
@@ -55,7 +55,7 @@
             = link_to image_tag("create-group.png"), [:new, :group], class: "create-group-btn"
           ul.groups
-            - @owner.groups.each do |group|
+            - @owner.groups.active.each do |group|
```
オーナー詳細画面左

before
![owners-show-before](https://user-images.githubusercontent.com/5820754/45884029-b36d8c80-bded-11e8-88b8-5f54ffe20e7c.png)

after
![owners-show-after](https://user-images.githubusercontent.com/5820754/45884046-bec0b800-bded-11e8-8ba5-b4dd7fcd7639.png)

noriyotcp commented on app/controllers/owners_controller.rb:
```diff
@@ -1,6 +1,6 @@
 class OwnersController < ApplicationController
   def index
-    @owners = User.order(name: :asc).all + Group.order(name: :asc).all
+    @owners = User.order(name: :asc).all + Group.active.order(name: :asc).all
```
collaboratorを選択する際に、削除されたグループを表示しないようにした

## before
![choose-a-collaborator-before](https://user-images.githubusercontent.com/5820754/45884111-e879df00-bded-11e8-9262-a45a75b2d0a0.png)

## after
![choose-a-collaborator-after](https://user-images.githubusercontent.com/5820754/45884120-f3347400-bded-11e8-9296-404b5495e8cb.png)

itkrt2y commented on app/controllers/groups_controller.rb:
```diff
@@ -1,20 +1,19 @@
   before_action :load_group, only: [:edit, :update, :destroy]
-
-  authorize_resource
+  before_action :forbidden, only: [:edit, :update, :destroy]
```
どちらかというと、ログイン画面にリダイレクトしてほしい

itkrt2y replied:
あ、違うか。current_userの権限を見ているのだから、そういう話じゃなかった

itkrt2y commented on spec/models/group_spec.rb:
```diff
@@ -30,4 +30,28 @@
+      FactoryBot.create_list(:group, 2, is_deleted: true)
+    end
+
+    it { is_expected.to eq Group.where(is_deleted: false) }
```
この書き方だと `create_list` で2個ずつレコード作った意味がないように思えますが、どのような理由でこのように書きましたか？
あと、このレベルだとダメというわけではないですが、テストの中にあまりロジックは書きたくないです。

noriyotcp replied:
返ってくるグループが「全て」`is_deleted: false` になっているというイメージで複数にしてましたが、false/true それぞれ1個ずつ作成して、返ってくるのは`false` のもの1個だけ、でも充分ですね

itkrt2y commented on spec/controllers/groups_controller_spec.rb:
```diff
@@ -106,34 +117,63 @@
+
+      describe 'ActiveRecord::RecordNotSaved' do
+        let(:error) { ActiveRecord::RecordNotSaved }
+        include_examples 'rescue'
```
なぜ `shared_examples` を使ってるんです？

noriyotcp replied:
もう一つ`ActiveRecord::RecordNotDestroyed` の`describe` があったけど、それは削除したので、`shared_examples` 使うまでもなくなりましたね 😩

itkrt2y commented on spec/controllers/groups_controller_spec.rb:
```diff
@@ -10,11 +10,22 @@
+
+      it 'fetches active projects only' do
+        subject
+        expect(assigns(:groups)).to be_all { |g| g.is_deleted == false }
```
`be_none { |group| group.is_deleted }` かな

itkrt2y commented on app/controllers/groups_controller.rb:
```diff
@@ -54,11 +55,13 @@ def additional_params
 
-    def load_group
-      @group = Group.friendly.find params[:id]
+    def forbidden
```
最近は一周回ってcancancanで `can?(:destroy, @group)` とかやればいいかなって気持ちになっていますが、秋田さんはどう思いますか？

itkrt2y replied:
こんなイメージ

```rb
def destroy
  @group = hogehoge
  if can?(:destroy, @group)
    @group.soft_destroy!
    redirect_to :groups
  else
    render_forbidden
  end
```

itkrt2y replied:
こんなイメージ

```rb
def destroy
  group = get_group
  if can?(:destroy, group)
    group.soft_destroy!
    redirect_to groups_path
  else
    render_forbidden
  end
```

noriyotcp replied:
cancancan は無くていきたい、という意向なのかなと思って、別の方法にしてました
上記のようなシンプルな使い方なら構わないと思います

itkrt2y replied:
ここはどうしますか？ このままでもいいですけど、個人的にはbefore_actionをやめたい

noriyotcp replied:
なるほどわかりました、`can?` でいきます

noriyotcp committed:
Modify specs

itkrt2y commented on spec/controllers/groups_controller_spec.rb:
```diff
@@ -106,34 +117,56 @@
+        allow_any_instance_of(Group).to receive(:soft_destroy!).and_raise(ActiveRecord::RecordNotSaved)
+      end
+
+      it { is_expected.to redirect_to [:edit, group] }
```
`[:edit, group]` って書き方、greppability低いですしroutesを変更した時にpathが存在しなくなってもエラーにならなかったり修正が漏れたり面倒なので、`edit_group_path` のように書いて欲しいです

itkrt2y commented on spec/models/group_spec.rb:
```diff
@@ -30,4 +30,28 @@
+      FactoryBot.create(:group, is_deleted: true)
+    end
+
+    it { is_expected.to be_none { |g| g.is_deleted } }
```
これ、例えば `Group.active` の結果が空配列でもテスト通ると思うので、素直なやり方じゃないですね。
ちゃんとどんな配列が得られるかでassertしないと。

```rb
let!(:active) { FactoryBot.create(:group, is_deleted: false) }
let!(:deleted) { FactoryBot.create(:group, is_deleted: true) }
it { is_expected.to eq [active] }
```

itkrt2y commented on app/views/groups/edit.html.slim:
```diff
@@ -23,7 +23,12 @@
+            'Delete this group
+          = link_to "Delete", group_path(@group), method: :delete, class: "btn", id: "delete-btn", data: { confirm: "Are you sure to delete this group?" }
+
+        - content_for :bottom
```
この行を元に戻す

noriyotcp replied:
確認ですが、この行というのは`- content_for :bottom` のみを元に戻すということでしょうか？

itkrt2y replied:
そうです。インデントが深くなってるけど、それ間違いですよね？

noriyotcp replied:
そうでした。修正しました 🙇

itkrt2y commented on spec/controllers/groups_controller_spec.rb:
```diff
@@ -10,11 +10,22 @@
+
+      it 'fetches active projects only' do
+        subject
+        expect(assigns(:groups)).to be_none { |g| g.is_deleted }
```
これもさっきのコメントと同様

noriyotcp committed:
Remove before_action

use load_group and can? in an each action instead

itkrt2y commented on app/controllers/groups_controller.rb:
```diff
@@ -54,11 +68,11 @@ def additional_params
-    def load_group
-      @group = Group.friendly.find params[:id]
+    def render_forbidden
+      render file: "public/403.html", status: :forbidden
```
ついでなんで `ApplicationController` に定義されてる `render_403` とか `render_40x` 系メソッドを整理して使ってもらえると、エラーハンドリングがDRYになっていいですね

itkrt2y replied:
今はどこにも使われてないメソッドで、多分そのままだと使えないので修正が必要

itkrt2y commented on app/controllers/groups_controller.rb:
```diff
@@ -54,11 +68,11 @@ def additional_params
-    def build_group
-      @group = Group.new(group_params)
+    def load_group
+      current_user.groups.active.friendly.find params[:id]
```
`load_group` って名前はcancancan向けのもので一般的には少々妙なネーミングであるのと、params内でのproject idのキーが必ず `:id` であると期待するのはよろしくないので、 `def get_group(id)` とか `def find_group(id)` とかにするべきでしょうね

itkrt2y replied:
paramsについてもう少し汎用的な考え方をすると、controllerではaction以外はparamsや他メソッドでセットされたインスタンス変数に依存するべきではなくて、メソッド内で使う値は全て引数で渡されるべき

noriyotcp committed:
Rename load_group to get_group(id)

pass params[:id]

noriyotcp committed:
Make render_403 public

noriyotcp commented on app/controllers/application_controller.rb:
```diff
@@ -26,6 +26,10 @@ def current_user=(user)
 
+  def render_forbidden(layout:)
+    render_403(layout: layout)
+  end
```
@itkrt2y
`render_403`のみ整理してみました
呼び出し時にlayoutの有無を指定するという形にしましたが、方向性としてはこれでいいのでしょうか？

itkrt2y replied:
違います

itkrt2y replied:
まず、Rubyはprivateメソッドを継承先でも呼べる言語なので、各Controllerで `render_403` を直接呼べばいい。というか、privateの必要性がないのでpublicにすればいい。
少なくとも、privateメソッドをpublicメソッドでラップするメリットはない

itkrt2y replied:
layoutの有無を引数で指定できるようにするのはいいですが、基本的な使い方を示すためにデフォルト引数は用意するべき

noriyotcp replied:
> 継承先でも呼べる

！！！ http://secret-garden.hatenablog.com/entry/2015/07/04/000000
確かにそもそもpublicでいいですね

itkrt2y replied:
これはApplicationControllerで定義しているものであるために汎用的に使われるメソッドになるので、汎用性を意識してみてください

noriyotcp replied:
`public/403.html` をレンダリングするということで`render_403` という名前をそのまま採用しました（statusは`:forbidden`のままですが）
一般的にはステータスコードでの表現のほうがいいのでしょうか 🤔

itkrt2y replied:
そこについては一般的なんてものはないので好きにすればいいです。僕だったらメソッド名を `forbidden` とか `unauthorized` とかにしますし。
ただ、メソッド名をこうした場合はHTTPステータスコードでしか覚えてない人にわかりづらいので、メソッド名を `def forbidden` にした上で `render file: "hoge",  status: 403` とrenderに渡すコードは数値にする。そうすれば、ステータスコード or ステータス名のどちらか一方しかわからなくてもgrepできる。
逆でもよくて、メソッド名は `render_403` で `render status: :forbidden` でもいい。そこは趣味だけど、greppabilityだけは意識すべき。

noriyotcp replied:
なるほど、その他render_40x についても同様に整理していきます

itkrt2y replied:
上のコメント、ちょっと追記しました

noriyotcp commented on app/controllers/application_controller.rb:
```diff
@@ -39,8 +43,8 @@ def render_401(_exception = nil)
-    def render_403(_exception = nil)
-      render file: Rails.root.join('public/403.html'), status: 403, layout: false, content_type: 'text/html'
+    def render_403(layout: false)
+      render file: Rails.root.join('public/403.html'), status: :forbidden, layout: layout, content_type: 'text/html'
```
`_exception = nil` は使われてない？ので削除
status は数字よりもシンボルのほうがわかりやすいかと思い変更

itkrt2y replied:
メソッド自体が使われてないので、引数以前の問題ですね

noriyotcp committed:
Make render_401 and render_404 public

itkrt2y committed:
Remove verbose content_type

itkrt2y committed:
Refactoring

itkrt2y commented:
方針はこれでOK。
デザインとか微修正したいところはありますが、ここでやりとりしても非効率的なので僕が巻き取るので、秋田さんは次のタスクをお願いします。

noriyotcp commented:
了解しました。よろしくお願いします。

itkrt2y committed:
Fix style

itkrt2y committed:
Restrict params correctly

itkrt2y committed:
Cache file on save failed
